cmd/specdiff/differ.go

```diff
@@ -0,0 +1,283 @@
+	switch changePath.changeType {
+	case "added":
+		diffProto.Added = append(diffProto.Added, fieldName)
+	case "deleted":
+		diffProto.Deleted = append(diffProto.Deleted, fieldName)
+	}
+}
+
+// getChanges creates a protodif report from a diff.Diff struct.
+func getChanges(diff *diff.Diff) (*rpc.Diff, error) {
+	diffProto := &rpc.Diff{
+		Added:        []string{},
+		Deleted:      []string{},
+		Modification: make(map[string]*rpc.Diff_ValueChange),
+	}
+	change := &change{
+		fieldPath:  stack{},
+		changeType: "",
+	}
+	diffNode := reflect.ValueOf(diff)
+	err := searchNode(diffNode, diffProto, change)
+	return diffProto, err
+}
+
+func searchNode(value reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	// Some values might be pointers, so we should dereference before continuing.
+	value = reflect.Indirect(value)
+
+	// Invalid values aren't relevant to the diff.
+	if !value.IsValid() {
+		return nil
+	}
+
+	switch value.Kind() {
+	case reflect.Map:
+		return searchMapType(value, diffProto, changePath)
+	case reflect.Array, reflect.Slice:
+		return searchArrayAndSliceType(value, diffProto, changePath)
+	case reflect.Struct:
+		return searchStructType(value, diffProto, changePath)
+	case reflect.Float64, reflect.String, reflect.Bool, reflect.Int:
+		valueString := getAtomicType(value)
+		changePath.fieldPath.push(valueString)
+		addToDiffProto(diffProto, changePath)
+		changePath.fieldPath.pop()
+		return nil
+	default:
+		return fmt.Errorf("field %q has unknown type %s with value %v", changePath.fieldPath, value.Type(), value)
+	}
+}
+
+func searchMapType(mapNode reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	if mapNode.Kind() != reflect.Map {
+		panic("searchMapType called with invalid type")
+	}
+
+	for _, childNodeKey := range mapNode.MapKeys() {
+		childNode := mapNode.MapIndex(childNodeKey)
+		if childNode.IsZero() {
+			continue
+		}
+
+		if isAtomicType(childNodeKey) {
+			childNodeKeyName := getAtomicType(childNodeKey)
+			changePath.fieldPath.push(childNodeKeyName)
+			err := searchNode(childNode, diffProto, changePath)
+			if err != nil {
+				return err
+			}
+			changePath.fieldPath.pop()
+			continue
+		}
+		if endpoint, ok := childNodeKey.Interface().(diff.Endpoint); ok {
+			changePath.fieldPath.push(handleEndpointStruct(endpoint))
+			err := searchNode(childNode, diffProto, changePath)
+			if err != nil {
+				return err
+			}
+			changePath.fieldPath.pop()
+			continue
+		}
+		return fmt.Errorf("map node key %v is not supported", childNodeKey)
+
+	}
+	return nil
+}
+
+func searchArrayAndSliceType(arrayNode reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	if arrayNode.Kind() != reflect.Slice && arrayNode.Kind() != reflect.Array {
+		panic("searchArrayAndSliceType called with invalid type")
+	}
+
+	for i := 0; i < (arrayNode.Len()); i++ {
+		childNode := arrayNode.Index(i)
+		if childNode.IsZero() {
+			continue
+		}
+		if childNode.Kind() == reflect.String {
```

Comment:
Similar to the type checking in searchMapType you might be able to type check this using a switch instead.